### DISCUSSION TITLE
store.subscribe does not work with selectors that use `get()` (I believe)

### DISCUSSION THREAD

**Discussion by jack3898** [pmndrs/zustand#2651] [Q&A]
I have followed the amazing guide by TkDodo to put zustand in a react context, to instance and scope it to a component tree and stop it being global.

https://tkdodo.eu/blog/zustand-and-react-context

This has been really helpful for a component I am writing. However, the store within the context is no longer inside of the React component lifecycle so (I believe) it relies on a custom subscriber implementation for higher-level components in the tree to subscribe to changes. This is where reusability of the store can happen with context.

See https://tkdodo.eu/blog/zustand-and-react-context#react-context

Here's what I did:

First I define a selector implementation:

```ts
export function zustandSelect<
    StoreValue,
    SelectorValue,
    HandlerValue extends (newSelected: SelectorValue) => void,
>(selector: (store: StoreValue) => SelectorValue, handler: HandlerValue) {
    return function processStoreChange(newState: StoreValue, prevState: StoreValue): void {
        const prevSelected = selector(prevState);
        const newSelected = selector(newState);

        if (!Object.is(prevSelected, newSelected)) {
            handler(newSelected);
        }
    };
}
```

Then a hook which is a simple useEffect wrapper consumes that utility:

```ts
import { useEffect } from "react";
import type { StoreApi } from "zustand/vanilla";
import { zustandSelect } from "../utils/zustandSelect";

export function useZustandSubscriber<Store, SelectorValue>(
    store: StoreApi<Store>,
    selector: (state: Store) => SelectorValue,
    subscriber?: (newState: SelectorValue) => void,
): void {
    useEffect(() => {
        if (!subscriber) {
            return undefined;
        }

        return store.subscribe(zustandSelect(selector, subscriber));
    }, [store, selector, subscriber]);
}
```

And here's how it's consumed:

`useZustandSubscriber(store, (state) => state.property, onPropertyChange);`

This works great, until I use it like this:

`useZustandSubscriber(store, (state) => state.getProperty(), onPropertyChange);`

The only difference between the former is the use of calling a getter rather than accessing a static property value directly. This never triggers the subscriber, and as such the state never changes.

I believe the `get()` method in the store's getter function is returning the same thing, which always passes the `Object.is()` check.

Any ideas what is going on? Have I overengineered (and it there a better way to do this?). I would like the subscriber to fire even with the getter when things have changed.

Thanks a lot!

**Comment by jack3898** [comment]
I knew posting a question here would magically bring the answer to my mind. 🤣

TL;DR I overengineered.

I should have just used `useStore` over my own implementation. Annoyingly, I knew about this hook but it never came to mind! Argh 😅

```ts
export function useZustandSubscriber<Store, SelectorValue>(
    store: StoreApi<Store>,
    selector: (state: Store) => SelectorValue,
    subscriber?: (newState: SelectorValue) => void,
): void {
    const selected = useStore(store, selector);

    useEffect(() => {
        subscriber?.(selected);
    }, [selected, subscriber]);
}
```
